app/src/main/kotlin/at/bitfire/davdroid/ui/PermissionsFragment.kt

```diff
@@ -155,75 +103,165 @@ class PermissionsFragment: Fragment() {
-                haveJtxPermissions.value = jtxPermissions
-                needJtxPermissions.value = jtxPermissions
-            } else {
-                haveJtxPermissions.value = null
-                needJtxPermissions.value = null
+        )
+    }
+}
+
+@Composable
+@OptIn(ExperimentalPermissionsApi::class)
+fun PermissionSwitchRow(
+    text: String,
+    permissions: List<String>?,
+    fontWeight: FontWeight = FontWeight.Normal
+) {
+    if (permissions != null) {
+        val state = rememberMultiplePermissionsState(permissions = permissions.toList())
+
+        PermissionSwitchRow(
+            text = text,
+            fontWeight = fontWeight,
+            allPermissionsGranted = state.allPermissionsGranted,
+            onLaunchRequest = state::launchMultiplePermissionRequest
+        )
+    }
+}
+
+@Composable
+fun PermissionSwitchRow(
+    text: String,
+    allPermissionsGranted: Boolean,
+    modifier: Modifier = Modifier,
+    fontWeight: FontWeight = FontWeight.Normal,
+    onLaunchRequest: () -> Unit
+) {
+    Row(
+        modifier = modifier,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Text(
+            text = text,
+            modifier = Modifier.weight(1f),
+            fontWeight = fontWeight
+        )
+        Switch(
+            checked = allPermissionsGranted,
+            enabled = !allPermissionsGranted,
+            onCheckedChange = { checked ->
+                if (checked) {
+                    onLaunchRequest()
+                }
             }
+        )
+    }
+}
+
+@Composable
+fun PermissionsCard(keepPermissions: Boolean?, onKeepPermissionsRequested: () -> Unit) {
+    val context = LocalContext.current
+    val pm = context.packageManager
 
-            // "all permissions" switch
-            val allPermissions = contactPermissions &&
-                    calendarPermissions &&
-                    notificationPermissions &&
-                    (!openTasksAvailable || openTasksPermissions == true) &&
-                    (!tasksOrgAvailable || tasksOrgPermissions == true) &&
-                    (!jtxAvailable || jtxPermissions == true)
-            haveAllPermissions.value = allPermissions
-            needAllPermissions.value = allPermissions
+    val openTasksAvailable = pm.resolveContentProvider(ProviderName.OpenTasks.authority, 0) != null
+    val tasksOrgAvailable = pm.resolveContentProvider(ProviderName.TasksOrg.authority, 0) != null
+    val jtxAvailable = pm.resolveContentProvider(ProviderName.JtxBoard.authority, 0) != null
+
+    CardWithImage(
+        title = stringResource(R.string.permissions_title),
+        message = stringResource(
+            R.string.permissions_text,
+            stringResource(R.string.app_name)
+        ),
+        image = painterResource(R.drawable.intro_permissions),
+        modifier = Modifier.padding(8.dp)
+    ) {
+        // TODO: keep permissions switch
```

Comment:
Is this done? In this case, please remove the TODO

Reply:
Yup, my bad